src/structarray.jl

```diff
@@ -102,10 +102,18 @@ end
 
 Base.similar(s::StructArray, sz::Base.DimOrInd...) = similar(s, Base.to_shape(sz))
 Base.similar(s::StructArray) = similar(s, Base.to_shape(axes(s)))
-function Base.similar(s::StructArray{T}, sz::Tuple) where {T}
+function Base.similar(s::StructArray{T,N,C}, ::Type{T}, sz::NTuple{N,Int64}) where {T, N, C<:Union{Tuple, NamedTuple}}
     StructArray{T}(map(typ -> similar(typ, sz), fieldarrays(s)))
 end
 
+function Base.similar(s::StructArray{T,N,C}, S::Type, sz::NTuple{N,Int64}) where {T, N, C<:Union{Tuple, NamedTuple}}
+    # If not specified, we don't really know what kind of array to use for each
+    # interior type, so we just pick the first one arbitrarily. If users need
+    # something else, they need to be more specific.
+    f1 = fieldarrays(s)[1]
+    buildfromschema(typ -> similar(f1, typ, sz), S)
```

Comment:
The following change allows types like `Float32` to be passed in for `S`.

```suggestion
    if isstructtype(S)
        return StructArrays.buildfromschema(typ -> similar(f1, typ, sz), S)
    else
        return similar(f1, S, sz)
    end
```